### ISSUE TITLE
Expose why jobs are in "pending"

### ISSUE THREAD

**Issue by wenottingham** [ansible/awx#1487]
##### ISSUE TYPE
 - Feature Idea

##### COMPONENT NAME
 - API
 - UI

##### SUMMARY

Jobs can be queued for a variety of reasons:

- waiting on dependent SCM update
- waiting on dependent inventory update
- blocked due to another instance of the same job
- blocked to another job which has the inventory locked (see above 'dependent inventory update')
- waiting on spare job running capacity
- pending because of no enabled instances in the instance group that is assigned to the job/project update/etc

This should be surfaced somehow, even if just in the UI tooltip/popover.

##### ENVIRONMENT
* AWX version: current

**Comment by ghost** [comment]
This would be really useful for troubleshooting, as I've been running into an issue where if the awx service is relaunched at an inopportune time, the new task host will never pick up jobs, but will show as active in the pool, causing everything to sit in pending indefinitely, with no obvious explanation.

I've not been able to narrow the exact cause, but I assume there must be some sort of locking that occurs in the database, and is never released because the process that would have released it dies.

**Comment by one-t** [comment]
As a follow-on to this, I think it would be very useful to expose the capacity impact of the job

**Comment by srgvg** [comment]
@Comradephate I'm experiencing this behaviour too (2.1). How do you work around or resolve this?

**Comment by benapetr** [comment]
+1

**Comment by kdelee** [comment]
Now that there are more scenarios in which nodes may become unavailable but could later have connection restored, and we are setting capacity based on the result of the health check, we probably want to be able to explain when jobs are in pending because there is no capacity in the instance group the job is trying to get launched into. That could be because instances are disabled, or because they are all failing / have never run their health check, hence have no capacity.

**Comment by AlanCoding** [comment]
@chrismeyersfsu didn't you do this? I know it shows it for lack of capacity. What's current status / objectives?

**Comment by chrismeyersfsu** [comment]
The want is for this information to be surfaced in the UI. To that end we should expose it via the API. I think @wenottingham initial list is pretty good. Users want to know the business logic that causes a job to block & when capacity is the reason for a job not running then what detailed capacity reason.

As @kdelee pointed out, we won't run a job if there isn't enough capacity. Why isn't there enough capacity?

* All possible instances that the job could run on are disabled
* Say there are 2 instances that a job could run on. `I1` has used all of it's capacity except for `1` unit, not enough to run the job. `I2` is disabled so it's capacity is considered 0. We wouldn't want the message to be `Not enough capacity found for Job`. Putting on my product manager hat: "Of the 2 candidate instances, I1 and I2, the job can not start on I1 because 1 capacity remaining isn't enough to run the job that requires 10 capacity units. I2 is considered to have 0 capacity because it is disabled".
* Many other reasons

We will have to be _really_ careful with performance on this. Let's say we have two nodes `N1` and `N2`. All jobs runs on `N2` and `N2` only. `N2` is flapping healthy and not-healthy i.e. 0 capacity and 1 capacity. Say you have 10,000 jobs in `pending`. The reason those 10,000 jobs aren't running changes with every health check.

**Comment by AlanCoding** [comment]
> `I1` has used all of it's capacity except for `1` unit, not enough to run the job. `I2` is disabled so it's capacity is considered 0.

I can't find it now, but wasn't there another issue suggesting that we should expose `preferred_instance_groups` in the API for jobs or job templates?

If we're talking about going beyond what's already been done, I don't think throwing more into a text field is going to cut it. Put yourself in the user's shoes. Why does the message mention anything about `I2`? What relationship does `I2` have to the current job, to what they're doing? The answer is that its instance group is in the preference list for that job, and the `I2` instance is a member of that instance group, so it searches in order of IG preference list, and then through the sublist of instances, ordered by remaining capacity. I don't think you want to explain that. I think some graphical capacity diagnostic could explain it in a way that user's wouldn't balk at.

**Comment by chrismeyersfsu** [comment]
I like your idea of helping the user understand why the job isn't scheduled for capacity reasons.

I'd still like some engineer/support level logging that captures the state of capacity when the decision is made. I want to make sure we don't run into "My job isn't being scheduled, and when I look at the capacity status it should be" <-- this could happen if/when we have a bug. Maybe we present the status of the capacity wrong, maybe the capacity is flapping & when observed there is capacity but when the decision is made there is 0 capacity.
